src/providers/hana/qgshanaconnection.cpp

```diff
@@ -0,0 +1,621 @@
+{
+  if ( layerProperty.geometryColName.isEmpty() )
+    return QgsWkbTypes::NoGeometry;
+
+  QgsWkbTypes::Type ret;
+
+  QString sql = QStringLiteral( "SELECT upper(%1.ST_GeometryType()) AS geom_type FROM %2.%3 "
+                                "WHERE %1 IS NOT NULL LIMIT %4" ).arg(
+                  QgsHanaUtils::quotedIdentifier( layerProperty.geometryColName ),
+                  QgsHanaUtils::quotedIdentifier( layerProperty.schemaName ),
+                  QgsHanaUtils::quotedIdentifier( layerProperty.tableName ),
+                  QString::number( GEOM_TYPE_SELECT_LIMIT ) );
+
+  try
+  {
+    StatementRef stmt = mConnection->createStatement();
+    ResultSetRef rsGeomType = stmt->executeQuery( reinterpret_cast<const char16_t *>( sql.unicode() ) );
+    QgsWkbTypes::Type geomType = QgsWkbTypes::Unknown, prevGeomType = QgsWkbTypes::Unknown;
+    while ( rsGeomType->next() )
+    {
+      geomType = QgsWkbTypes::singleType( QgsHanaUtils::toWkbType( rsGeomType->getString( 1 )->c_str() ) );
+      if ( prevGeomType != QgsWkbTypes::Unknown && geomType != prevGeomType )
+      {
+        geomType = QgsWkbTypes::Unknown;
+        break;
+      }
+      prevGeomType = geomType;
+    }
+    ret = geomType;
+    rsGeomType->close();
+  }
+  catch ( const Exception &ex )
+  {
+    throw QgsHanaException( ex.what() );
+  }
+  return ret;
+}
+
+QString QgsHanaConnection::getColumnDataType( const QString &schemaName, const QString &tableName, const QString &columnName )
+{
+  QString sql = QStringLiteral( "SELECT DATA_TYPE_NAME FROM SYS.TABLE_COLUMNS WHERE SCHEMA_NAME = '%1' AND "
+                                "TABLE_NAME = '%2' AND COLUMN_NAME = '%3'" ).arg( schemaName, tableName, columnName );
+
+  QString ret;
+  try
+  {
+    StatementRef stmt = mConnection->createStatement();
+    ResultSetRef rsDataType = stmt->executeQuery( reinterpret_cast<const char16_t *>( sql.unicode() ) );
+    while ( rsDataType->next() )
+    {
+      ret = QgsHanaUtils::toQString( rsDataType->getString( 1 ) );
+    }
+    rsDataType->close();
+  }
+  catch ( const Exception &ex )
+  {
+    throw QgsHanaException( ex.what() );
+  }
+
+  return ret;
+}
+
+QgsHanaConnectionRef::QgsHanaConnectionRef( const QgsDataSourceUri &uri )
+{
+  mConnection = std::unique_ptr<QgsHanaConnection>(
+                  QgsHanaConnectionPool::instance()->acquireConnection( QgsHanaUtils::connectionInfo( uri ) ) );
+}
+
+QgsHanaConnectionRef::QgsHanaConnectionRef( const QString &name )
+{
+  QgsHanaSettings settings( name, true );
+  mConnection = std::unique_ptr<QgsHanaConnection>(
+                  QgsHanaConnectionPool::instance()->acquireConnection( QgsHanaUtils::connectionInfo( settings.toDataSourceUri() ) ) );
+}
+
+QgsHanaConnectionRef::~QgsHanaConnectionRef()
+{
+  if ( mConnection && QgsHanaConnectionPool::hasInstance() )
+    QgsHanaConnectionPool::instance()->releaseConnection( mConnection.release() );
```

Comment:
Even if `hasInstance()` returns `true`, another thread could delete the instance and then it would be created again by the next call to `QgsHanaConnectionPool::instance()`.

Even worse: `hasInstance()` returns `true`, `instance()` returns a pointer to the instance, then it's deleted by another thread before or while this thread calls `releaseConnection()`.

Reply:
Done in 97d98ca.